lib/openstack_query/runners/query_runner.py

```diff
@@ -23,50 +24,49 @@ def __init__(self, connection_cls=OpenstackConnection):
     def run(
         self,
         cloud_account: str,
-        filter_func: Optional[ParsedFilterFunc] = None,
-        filter_kwargs: Optional[ServerSideFilters] = None,
+        client_side_filter_func: Optional[ClientSideFilterFunc] = None,
+        server_side_filters: Optional[ServerSideFilters] = None,
         from_subset: Optional[List[Any]] = None,
         **kwargs
     ) -> List[OpenstackResourceObj]:
         """
         Public method that runs the query by querying openstacksdk and then applying a filter function.
         :param cloud_account: The account from the clouds configuration to use
-        :param filter_func: An Optional function that we can use to limit the results after querying openstacksdk
-        :param filter_kwargs: An Optional set of filter kwargs to limit the results by when querying openstacksdk
+        :param client_side_filter_func: An Optional function that we can use to limit the results after querying openstacksdk
+        :param server_side_filters: An Optional set of filter kwargs to limit the results by when querying openstacksdk
         :param from_subset: A subset of openstack resources to run query on instead of querying openstacksdk
         :param kwargs: An extra set of kwargs to pass to internal _run_query method that changes what/how the
         openstacksdk query is run
             - valid kwargs to _run_query is specific to the runner object - see docstrings for _run_query() on the
             runner of interest.
         """
 
-        force_filter_func_usage = False
+        if from_subset and server_side_filters:
+            raise ParseQueryError("Cannot parse from a subset and server filter")
+
         with self._connection_cls(cloud_account) as conn:
-            if from_subset:
-                results_resource_objects = self._parse_subset(conn, from_subset)
-                force_filter_func_usage = True
-            else:
-                results_resource_objects = self._run_query(
-                    conn, filter_kwargs, **kwargs
-                )
+            resource_objects = (
+                self._parse_subset(conn, from_subset)
+                if from_subset
+                else self._run_query(conn, server_side_filters, **kwargs)
+            )
 
-        if filter_kwargs is None or force_filter_func_usage:
-            results_resource_objects = self._apply_filter_func(
-                results_resource_objects,
-                filter_func,
+        if client_side_filter_func and not server_side_filters:
+            resource_objects = self._apply_client_side_filter(
+                resource_objects, client_side_filter_func
```

Comment:
Any reason we can't have it do a client side after a server side?

Reply:
Well - right now we allow only one Preset per query - each preset has:
- an associated `client_side_filter_func`
- an optional set of `server_side_filters` 
They both do the same filtering - hence we only need one or the other

Reply:
Based on our discussion in person: definitely for a future PR but I can see us only allowing a single server side filter with n client side